pkg/graveler/committed/prefix_iterator.go

```diff
@@ -0,0 +1,174 @@
+			ipi.updatePath()
+		}
+	} else {
+		for vr != nil && ipi.position != done && strings.HasPrefix(vr.Key.String(), string(ipi.prefixes[ipi.position])) {
+			if ipi.rangeIterator.Err() != nil {
+				return false
+			}
+			ipi.hasNext = ipi.rangeIterator.Next()
+			if !ipi.hasNext {
+				return false
+			}
+			vr, _ = ipi.updateValue()
+			ipi.updatePath()
+		}
+	}
+	return true
+}
+
+func (ipi *PrefixIterator) NextRange() bool {
+	ipi.hasNext = ipi.rangeIterator.NextRange()
+	if !ipi.hasNext {
+		return false
+	}
+	ipi.updateValue()
+	ipi.updatePath()
+
+	for ipi.IsCurrentRangeBoundedByPath() {
+		if ipi.rangeIterator.Err() != nil {
+			return false
+		}
+		ipi.hasNext = ipi.rangeIterator.NextRange()
+		if !ipi.hasNext {
+			return false
+		}
+		ipi.updateValue()
+		ipi.position++
+		if ipi.position >= len(ipi.prefixes) {
+			ipi.position = done
+		}
+	}
+	return true
+}
+
+func NewPrefixIterator(prefixes []graveler.Prefix, rangeIterator Iterator) *PrefixIterator {
+	sort.Slice(prefixes, func(i, j int) bool {
+		return prefixes[i] < prefixes[j]
+	})
+	return &PrefixIterator{prefixes: prefixes, position: 0, rangeIterator: rangeIterator}
+}
+
+func (ipi *PrefixIterator) updatePath() {
+	if ipi.position == done {
+		return
+	}
+	currMinKey := string(ipi.currentRangeValue.r.MinKey)
+	if ipi.currentRangeValue.vr != nil {
+		currMinKey = string(ipi.currentRangeValue.vr.Key)
+	}
+	// If the position is smaller or doesn't have the prefix of the current key, get the next prefix.
+	// At the end of this function we'll have either a path that is a prefix of the current key, or bigger than the key
+	for string(ipi.prefixes[ipi.position]) < currMinKey &&
+		!strings.HasPrefix(currMinKey, string(ipi.prefixes[ipi.position])) {
+		p := ipi.position + 1
+		switch {
+		case p >= len(ipi.prefixes): // No more comparable prefixes
+			ipi.position = done
+			return
+		case string(ipi.prefixes[p]) <= string(ipi.currentRangeValue.r.MaxKey):
+			ipi.position = p
+		default:
+			return
+		}
+	}
+}
+
+func (ipi *PrefixIterator) getPrefix() *graveler.Prefix {
+	if ipi.position == done {
+		return nil
+	}
+	return &ipi.prefixes[ipi.position]
+}
+
+// IsCurrentRangeBoundedByPath returns true if both the range's max and min keys have the current path as their prefix
+func (ipi *PrefixIterator) IsCurrentRangeBoundedByPath() bool {
+	p := ipi.getPrefix()
+	if p == nil {
+		return false
+	}
+	r := ipi.currentRangeValue.r
+	return strings.HasPrefix(string(r.MinKey), string(*p)) && strings.HasPrefix(string(r.MaxKey), string(*p))
+}
+
+// IsCurrentPathIncludedInRange returns true if the examined path is either a prefix of the range's min or max key,
+// or if the path is between the range's min and max keys.
+func (ipi *PrefixIterator) IsCurrentPathIncludedInRange() bool {
+	p := ipi.getPrefix()
+	if p == nil {
+		return false
+	}
+	r := ipi.currentRangeValue.r
+	hasPrefix := strings.HasPrefix(string(r.MinKey), string(*p)) || strings.HasPrefix(string(r.MaxKey), string(*p))
```

Comment:
> if the range's max key has the prefix, but the prefix does not intersect with the range

Don't understand this line, how is it possible that the range's max key has the prefix and it doesn't intersect the range?

Prefix: `c` 
Range: `[a1, c6, d7]`
The prefix intersects the range yet not the prefix of the min or max key.

Reply:
You're right, I didn't explain it correctly. I still think the second condition may be redundant - edited the original comment.

Reply:
Still not sure how the condition is redundant...
If the range's max key has the prefix, then the first condition meets.
If the range's max (or min) key doesn't have the prefix, it might be between them, for example the comment above.